docs/development/design/explicit-builders.rst

```diff
@@ -0,0 +1,178 @@
+
+The new builders would do the minimal (or none) changes to the user's configuration in order to build their docs.
+
+Sphinx
+~~~~~~
+
+conf.py
+'''''''
+
+We should read the configuration file from the user or default to *one* path,
+and error if it doesn't exist.
+We shouldn't change or override the values from the user's configuration (``conf.py``),
+some are:
+
+- ``_static`` is added to ``html_static_path`` (probably an old setting)
+- ``html_theme`` (we are only setting this if certain condition is met)
+- ``websupport2_*`` (probably old settings)
+- ``html_context`` (more information below)
+- Latex settings
+  (they are used to improve the output for Chinese and Japanese languages, we should write a guide instead)
+
+Sphinx's ``html_context``
+'''''''''''''''''''''''''
+
+We should pass the minimal needed information into the context.
+This is related to :doc:`/development/design/theme-context`.
+
+With :doc:`/api/v3` and environment variables (to store the secret token)
+should be possible to query several things from the project without having to pass them at build time into the context.
+Most the of basic information can be obtained from our environment variables (``READTHEDOCS_*``).
+
+Some values from the context are used in our Sphinx extension,
+we should define them as configuration options instead.
+Others are used in our theme, should we stop setting them?
+
+Extension
+'''''''''
+
+There are some things our extension does that are already supported by Sphinx or themes
+(analytics, canonical URL, etc), we should write a guide instead of applying our magic.
+
+The extension also injects some custom js and css files.
+We could try another more general approach and inject these on each HTML file after the build is complete.
+
+The extension injects the warning banner if you are reading the docs from a pull request.
+We could implement this like the version warning banner instead, so it works for MkDocs too.
+(this is injected with the version selector).
+
+MkDocs
+~~~~~~
+
+We should read the configuration file from the user or default to *one* path,
+and error if it doesn't exist.
+
+We shouldn't change the values from the user's configuration (``mkdocs.yml``),
+currently we change:
+
+- ``docs_dir``
+- ``extra_javascript``
+- ``extra_css``
+- ``google_analytics`` (we change this to ``None`` and use our own method)
+- ``theme`` (we set it to our theme for old projects)
+
+Only the additional js/css files should be added.
+Additionally, we could try another more general approach and inject these after the build is complete.
+
+Related to `#7844`_, `#4924`_, `#4827`_, `#4820`_
+
+.. _#7844: https://github.com/readthedocs/readthedocs.org/issues/7844
+.. _#4924: https://github.com/readthedocs/readthedocs.org/issues/4924
+.. _#4827: https://github.com/readthedocs/readthedocs.org/issues/4827
+.. _#4820: https://github.com/readthedocs/readthedocs.org/issues/4820
+
+Web settings
```

Comment:
Should we get rid of web settings entirely for things that are in the config file?

Reply:
I think they are still useful when you want to build from an existing release, since you can't change that code to add a config file.

Reply:
I think we should remove them and provide a nice way to support version aliases. So, in case you have to re-build an old tag (e.g. `2.1`) you can create a new branch/tag (e.g.`2.1.1`) and keep the same URL and everything but with updated/fixed docs by using an alias for that new branch/tag.